<a href="https://colab.research.google.com/github/LitvinovaYaI/ML_Credit_cards/blob/main/ML_task2_part2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.svm import LinearSVC
from sklearn.metrics import classification_report
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score
from sklearn.ensemble import RandomForestClassifier
from keras.models import Sequential
from keras.layers import InputLayer, Dense, BatchNormalization
from keras.callbacks import ModelCheckpoint

In [4]:
data = pd.read_csv("/content/creditcard.csv")
data.head()
data.info()
data = data.drop_duplicates(keep='first')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [5]:
impostor = data[data['Class']==1]
legal = data[data['Class']==0]
legal_sampled = data[data['Class']==0].sample(n = impostor.shape[0], random_state=21)
print(f' Impostor Shape{impostor.shape}\n Legal shape{legal_sampled.shape}')
impostor_sampled = impostor
final_data = pd.concat([impostor_sampled, legal_sampled], ignore_index=True)
final_data.shape

 Impostor Shape(473, 31)
 Legal shape(473, 31)


(946, 31)

In [8]:
x_dropped=data.drop(data.columns[[0,1,2,3,4,5,6,7,8,9,10,11,13,15,16,20,21,22,23,24,25,26,27,28,29,30]], axis=1)
print(x_dropped)
scaler=StandardScaler()
x_dropped = scaler.fit_transform(x_dropped)
y=data['Class']
x_dropped, x_dropped_test, y_dropped, y_dropped_test = train_test_split(x_dropped,y, test_size=0.20, random_state=21)
print(f'x_train{x_dropped.shape}\n, x_test{x_dropped_test.shape}\n, y_train{y_dropped.shape}\n, y_test{y_dropped_test.shape}')

             V12       V14       V17       V18       V19
0      -0.617801 -0.311169  0.207971  0.025791  0.403993
1       1.065235 -0.143772 -0.114805 -0.183361 -0.145783
2       0.066084 -0.165946  1.109969 -0.121359 -2.261857
3       0.178228 -0.287924 -0.684093  1.965775 -1.232622
4       0.538196 -1.119670 -0.237033 -0.038195  0.803487
...          ...       ...       ...       ...       ...
284802  2.711941  4.626942  1.991691  0.510632 -0.682920
284803  0.915802 -0.675143 -0.025693 -1.221179 -1.545556
284804  0.063119 -0.510602  0.313502  0.395652 -0.577252
284805 -0.962886  0.449624  0.509928  1.113981  2.897849
284806 -0.031513 -0.084316 -0.660377  0.167430 -0.256117

[283726 rows x 5 columns]
x_train(226980, 5)
, x_test(56746, 5)
, y_train(226980,)
, y_test(56746,)


In [12]:
# Logistic Regression model
def logReg(train_x, train_y, x_test, y_test):
    logistic_model = LogisticRegression()
    logistic_model.fit(train_x, train_y)
    y_pred_train = logistic_model.predict(train_x)
    y_pred_test = logistic_model.predict(x_test)
    res_train = classification_report(train_y, y_pred_train, target_names=["Imposter","Legal"])
    res_test = classification_report(y_test, y_pred_test, target_names=["Imposter","Legal"])
    print(res_train)
    print("*"*100)
    print(res_test)

In [10]:
#KNeighbors classification model
def KNeighbors(x_train, y_train, x_test, y_test):
  Kneib = KNeighborsClassifier(n_neighbors=3)
  Kneib.fit(x_train, y_train)
  y_pred_train = Kneib.predict(x_train)
  y_pred_test = Kneib.predict(x_test)
  res_train = classification_report(y_train, y_pred_train, target_names=["Imposter","Legal"])
  res_test = classification_report(y_test, y_pred_test, target_names=["Imposter","Legal"])
  print(res_train)
  print("*"*100)
  print(res_test)

In [13]:
LG_dropped = logReg(x_dropped, y_dropped, x_dropped_test, y_dropped_test)

              precision    recall  f1-score   support

    Imposter       1.00      1.00      1.00    226599
       Legal       0.86      0.57      0.68       381

    accuracy                           1.00    226980
   macro avg       0.93      0.78      0.84    226980
weighted avg       1.00      1.00      1.00    226980

****************************************************************************************************
              precision    recall  f1-score   support

    Imposter       1.00      1.00      1.00     56654
       Legal       0.86      0.52      0.65        92

    accuracy                           1.00     56746
   macro avg       0.93      0.76      0.82     56746
weighted avg       1.00      1.00      1.00     56746



In [14]:
KN_dropped = KNeighbors(x_dropped, y_dropped, x_dropped_test, y_dropped_test)

              precision    recall  f1-score   support

    Imposter       1.00      1.00      1.00    226599
       Legal       0.94      0.81      0.87       381

    accuracy                           1.00    226980
   macro avg       0.97      0.91      0.94    226980
weighted avg       1.00      1.00      1.00    226980

****************************************************************************************************
              precision    recall  f1-score   support

    Imposter       1.00      1.00      1.00     56654
       Legal       0.91      0.66      0.77        92

    accuracy                           1.00     56746
   macro avg       0.95      0.83      0.88     56746
weighted avg       1.00      1.00      1.00     56746



Итак, результаты для полного датасета: метод линейной регрессии 93%, метод ближайших соседей 97%, для обрезанного датасета: метод линейной регрессии 93%, метод ближайших соседей 95%. Сохранение только наиболее коррелируемых параметров не улучшило результат.